File: C/SCLL.c
/////////////////////////////////////////////////////////////////////////////////////////////////////
//                                                                                                 //
//  File name :     SCLL.c                                                                         //
//  Description :   Implementation of a Singly Circular Linked List (SCL) in C.                    //
//  Author :        Pallavi Omprakash Malewar                                                      //
//  Date :          27/10/2025                                                                     //
//                                                                                                 //
/////////////////////////////////////////////////////////////////////////////////////////////////////


#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

int Count(PNODE first, PNODE last)
{
    int iCount = 0;

    if(first == NULL)
    {
        return iCount;
    }

    do
    {
        iCount++;
        first = first -> next;
    } while(first != last -> next);
    
    return iCount;
}

void InsertFirst(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    if(newn == NULL)
    {
        printf("Memory Allocation is Failed.\n");
        return;
    }

    newn -> data = iNo;
    newn -> next = NULL;

    if((*first) == NULL && *last == NULL)
    {
        (*first) = newn;
        *last = newn;

        (*last) -> next = (*first);
    }
    else
    {
        newn -> next = (*first);
        (*first) = newn;

        (*last) -> next = *first;
    }
}

void InsertLast(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    if(newn == NULL)
    {
        printf("Memory Allocation is Failed.\n");
        return;
    }

    newn -> data = iNo;
    newn -> next = NULL;

    if(*first == NULL && *last == NULL)
    {
        *first = newn;
        *last = newn;

        (*last) -> next = (*first);
    }
    else
    {
        (*last) -> next = newn;
        *last = newn;

        (*last) -> next = *first;
    }

}

void InsertAtPos(PPNODE first, PPNODE last, int iNo, int iPos)
{
    int iCount = 0;
    iCount = Count(*first, *last);

    if((iPos <= 0) || (iPos > iCount + 1))
    {
        printf("Invalid Position\nPlease Enter Valid Input.\n");
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(first, last, iNo);
    }
    else if(iPos == iCount+1)
    {
        InsertLast(first, last, iNo);
    }
    else
    {
        PNODE newn = NULL;

        newn = (PNODE)malloc(sizeof(NODE));

        if(newn == NULL)
        {
            printf("Memory Allocation is Failed.\n");
            return;
        }

        newn -> data = iNo;
        newn -> next = NULL;

        PNODE temp = NULL;

        temp = (*first);

        int i = 0;

        for(i = 1 ; i < iPos - 1 ; i++)
        {
            temp = temp -> next;
        }

        newn -> next = temp  -> next;
        temp -> next = newn;
    }

}

void DeleteFirst(PPNODE first, PPNODE last)
{
    if((*first) == NULL && (*last) == NULL)
    {
        printf("Linked List is Empty.\n");
        return;
    }

    else if(*first == *last)
    {
        free(*first);
        (*first) = NULL;
        (*last) = NULL;
    }
    else
    {
        PNODE temp = (*first);
        *first = (*first) -> next;
        free(temp);

        (*last) -> next = *first;
    }
}

void DeleteLast(PPNODE first, PPNODE last)
{
    if(*first == NULL)
    {
        return;
    }
    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }
    else
    {
        PNODE temp = NULL;

        temp = (*first);

        while(temp -> next != (*last))
        {
            temp = temp -> next;
        }
        free(temp -> next);
        temp -> next = (*first);
        (*last) = temp;
    }
}

void DeleteAtPos(PPNODE first, PPNODE last, int iPos)
{
    int iCount = 0, i = 0;

    iCount = Count(*first, *last);

    if((iPos <= 0) || (iPos > iCount))
    {
        printf("Invalid Position.\n");
        return;
    }

    if(iPos == 1)
    {
        DeleteFirst(first, last);
    }
    else if(iPos == iCount)
    {
        DeleteLast(first, last);
    }
    else
    {
        PNODE temp = NULL;
        PNODE target = NULL;

        temp = (*first);

        for(i = 1 ; i < iPos -1; i++)
        {
            temp = temp -> next;
        }
        target = temp -> next;

        temp -> next = target -> next;
        free(target);   
    }
}

void Display(PNODE first, PNODE last)
{
    if(first == NULL)
    {
        printf("List is Empty.\n");
        return;
    }

    PNODE temp = first;

    do
    {
        printf("| %d | -> ",temp -> data);
        temp = temp -> next;
    } while (temp != first);
    printf("\n");    
}

void DeleteAll(PPNODE first, PPNODE last)
{
    if(*first == NULL)
    {
        return;
    }    

    PNODE start = *first;
    PNODE temp = *first;

    PNODE nextNode = NULL;

    do
    {
        nextNode = temp -> next;
        free(temp);
        temp = nextNode;
    } while (temp != start);
    
    (*first) = NULL;
    (*last) = NULL;
}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;

    int iChoice = 0;
    int iValue = 0;
    int iPos = 0;
    int iRet = 0;

    printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");
    printf("---------------------------------------------------------Singly Circular Linked List-------------------------------------------------------------\n");
    printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");

    while(1)
    {
        printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");
        printf("---------------------------------------------------------Please Select the option--------------------------------------------------------------\n");
        printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");

        printf("1. Insert New Node at first Position.\n");
        printf("2. Insert New Node at last Position.\n");
        printf("3. Insert New Node at given Position.\n");
        printf("4. Delete the Node from first Position.\n");
        printf("5. Delete the Node from last Position.\n");
        printf("6. Delete the Node from given Position.\n");
        printf("7. Display all nodes of Linked List.\n");
        printf("8. Count Number of Nodes from Linked List.\n");
        printf("9. Delete the Entire Linked List.\n");
        printf("0. Terminate the Application.\n");

        printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");

        scanf("%d", &iChoice);

        if(iChoice == 1)
        {
            printf("Enter the data you want to insert : ");
            scanf("%d",&iValue);

            InsertFirst(&head, &tail, iValue);
        }
        else if(iChoice == 2)
        {
            printf("Enter the data you want to insert : ");
            scanf("%d",&iValue);

            InsertLast(&head, &tail, iValue);
        }
        else if(iChoice == 3)
        {
            printf("Enter the data you want to insert : ");
            scanf("%d", &iValue);

            printf("Enter the Position where you want to enter data : ");
            scanf("%d", &iPos);

            InsertAtPos(&head, &tail, iValue, iPos);
        }
        else if(iChoice == 4)
        {
            printf("Deleting the First element from Linked List.\n");

            DeleteFirst(&head, &tail);
        }
        else if(iChoice == 5)
        {
            printf("Deleting the Last element from Linked List.\n");

            DeleteLast(&head,  &tail);
        }
        else if(iChoice == 6)
        {
            printf("Enter the Position from which you want to delete the node : \n");
            scanf("%d", &iPos);

            printf("Deleting the node from Position : %d\n", iPos);

            DeleteAtPos(&head, &tail, iPos);
        }
        else if(iChoice == 7)
        {
            printf("Elements of the Linked List are :\n");
            Display(head, tail);
        }
        else if(iChoice == 8)
        {
            iRet = Count(head, tail);
            printf("Number of Nodes in the Linked List : %d\n", iRet);
        }
        else if(iChoice == 9)
        {
            DeleteAll(&head, &tail);
            printf("Entire Linked List is Deleted.\n");
        }
        else if(iChoice == 0)
        {
            DeleteAll(&head, &tail);
            printf("Thank You for using our Application.\n");
            break;
        }
        else
        {
            printf("Invalid Choice\n");
        }
        printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");
    }

    return 0;
}

File: JAVA/SCLL.java
///////////////////////////////////////////////////////////
//                                                       //
//  File name :     SCLL.cpp                             //
//  Description :   Singly Circular Linked List          //
//  Author :        Pallavi Omprakash Malewar            //
//  Date :          07/02/2026                           //
//                                                       //
///////////////////////////////////////////////////////////

import java.util.*;

class Node
{
    public int data;
    public Node next;
}

class SinglyCLL 
{
    private Node first;
    private Node last;
    private int iCount;

    public SinglyCLL()
    {
        first = null;
        last = null;

        iCount = 0;
    }

    public void InsertFirst(int iNo)
    {
        Node newn = null;

        newn = new Node();

        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
            last = newn;

            last.next = first;
        }
        else
        {
            newn.next = first;
            first = newn;
            last.next = newn;
        }
        iCount++;
    }

    public void InsertLast(int iNo)
    {
        Node newn = null;
        newn = new Node();
        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
            last = newn;

            last.next = first;
        }
        else
        {
            last.next = newn;
            last = newn;
            last.next = first;
        }
        iCount++;
    }

    public void InsertAtPos(int iNo, int iPos)
    {
        if(iPos < 1 || iPos > iCount+1)
        {
            System.out.println("Invalid Position.");
            return;
        }
        if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else if(iPos == iCount+1)
        {
            InsertLast(iNo);
        }
        else
        {
            Node temp = null;
            Node newn = null;

            int iCnt = 0;

            newn = new Node();

            newn.data = iNo;
            newn.next = null;

            temp = first;

            for(iCnt = 1 ; iCnt < iPos - 1 ; iCnt++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if(first == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;
            last.next = first;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        Node temp = null;

        if(first == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            temp = first;

            while(temp.next != last)
            {
                temp = temp.next;
            }
            last = temp;
            last.next = first;
        }
        iCount--;
    }

    public void DeleteAtPos(int iPos)
    {
        if(iPos < 1 || iPos > iCount)
        {
            System.out.println("Invalid Position!");
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst(); 
        }
        else if(iPos == iCount)
        {
            DeleteLast();
        }
        else
        {
            Node temp = null;
            int iCnt = 0;

            temp = first;

            for(iCnt = 1; iCnt < iPos - 1 ; iCnt++)
            {
                temp = temp.next;
            }

            temp.next = temp.next.next;
            iCount--;
        }
    }

    public void Display()
    {
        if(first == null)
        {
            System.out.println("Linked List is empty.");
            return;
        }

        Node temp = null;
        temp = first;

        do
        {
            System.out.print("| " + temp.data + " | -> ");
            temp = temp.next;
        }
        while(temp != first);
        
        System.out.println("..back to First..");
    }

    public int Count()
    {
        return iCount;
    }
    
    public void DeleteAll()
    {
        first = null;
        last = null;
        iCount = 0;

        System.out.println("Entire LinkedList Deleted.");
    }
}

class SCLL
{
    public static void main(String A[])
    {
        Scanner scobj = new Scanner(System.in);
        SinglyCLL sobj = new SinglyCLL();

        int iRet = 0;
        int iPos = 0;
        int iChoice = 0;
        int iValue = 0;


        System.out.println("-----------------------------------------------------------------------------------------------------------------------------------------------");
        System.out.println("---------------------------------------------------------Singly Circular Linked List-----------------------------------------------------------");
        System.out.println("-----------------------------------------------------------------------------------------------------------------------------------------------");

        while(true)
        {
            System.out.println("-----------------------------------------------------------------------------------------------------------------------------------------------");
            System.out.println("---------------------------------------------------------Please Select the option--------------------------------------------------------------");
            System.out.println("-----------------------------------------------------------------------------------------------------------------------------------------------");
           
            System.out.println("1. Insert New Node at first Position.");
            System.out.println("2. Insert New Node at last Position.");
            System.out.println("3. Insert New Node at given Position.");
            System.out.println("4. Delete the Node from first Position.");
            System.out.println("5. Delete the Node from last Position.");
            System.out.println("6. Delete the Node from given Position.");
            System.out.println("7. Display all nodes of Linked List.");
            System.out.println("8. Count Number of Nodes from Linked List.");
            System.out.println("9. Delete Entire Linked List.");
            System.out.println("0. Terminate the Application.");

            System.out.println("-----------------------------------------------------------------------------------------------------------------------------------------------");

            iChoice = scobj.nextInt();

            if(iChoice == 1)
            {
                System.out.println("Enter the data you want to insert: ");
                iValue = scobj.nextInt();

                sobj.InsertFirst(iValue);
            }
            else if(iChoice == 2)
            {
                System.out.println("Enter the data you want to insert: ");
                iValue = scobj.nextInt();

                sobj.InsertLast(iValue);
            }
            else if(iChoice == 3)
            {
                System.out.println("Enter the data you want to insert: ");
                iValue = scobj.nextInt();

                System.out.println("Enter the position where you want to enter data: ");
                iPos = scobj.nextInt();

                sobj.InsertAtPos(iValue, iPos);
            }
            else if(iChoice == 4)
            {
                System.out.println("Deleting the First Element from Linked List.");
                sobj.DeleteFirst();
            }
            else if(iChoice == 5)
            {
                System.out.println("Deleting the last Element from Linked List.");
                sobj.DeleteLast();
            }
            else if(iChoice == 6)
            {
                System.out.println("Enter the Position to delete data at that position: ");
                iPos = scobj.nextInt();

                System.out.println("Deleting the node from Position " + iPos);

                sobj.DeleteAtPos(iPos);
            }
            else if(iChoice == 7)
            {
                System.out.println("Elements of the Linked List are : ");
                sobj.Display();
            }
            else if(iChoice == 8)
            {
                iRet = sobj.Count();
                System.out.println("Number of Nodes in the Linked List : " + iRet);
            }
            else if(iChoice == 9)
            {
                sobj.DeleteAll();
            }
            else if(iChoice == 0)
            {
                System.out.println("Thank You for using our Appliaction.");
                break;
            }
            else
            {
                System.out.println("Invalid Choice.");
            }
            System.out.println("-----------------------------------------------------------------------------------------------------------------------------------------------");
        }

        scobj.close();
    }
}